app/scripts/controllers/user-storage/encryption.ts

```diff
@@ -23,14 +36,57 @@ function bytesToUtf8(byteArray: Uint8Array) {
   return decoder.decode(byteArray);
 }
 
+/**
+ * In Memory Cache of the derived key from a given salt.
+ * We can tidy this to make it reusable with future versions and KDF functions
+ */
+const inMemCachedKDF: Map<string, Uint8Array> = new Map();
+
+type CachedEntry = {
+  salt: Uint8Array;
+  key: Uint8Array;
+};
+const getCachedKeyBySalt = (salt: Uint8Array): CachedEntry | undefined => {
+  const base64Salt = byteArrayToBase64(salt);
+  const cachedKey = inMemCachedKDF.get(base64Salt);
+  if (!cachedKey) {
+    return undefined;
+  }
+
+  return {
+    salt,
+    key: cachedKey,
+  };
+};
+const getAnyCachedKey = (): CachedEntry | undefined => {
+  const cachedEntry = Array.from(inMemCachedKDF.entries()).at(0);
```

Comment:
(nitpick)
You can avoid iterating all entries to create array

```suggestion
  const cachedEntry = inMemCachedKDF.entries().next().value
```